go-controller/pkg/clustermanager/userdefinednetwork/controller_test.go

```diff
@@ -31,34 +33,46 @@ import (
 	udninformerfactory "github.com/ovn-org/ovn-kubernetes/go-controller/pkg/crd/userdefinednetwork/v1/apis/informers/externalversions"
 	udninformer "github.com/ovn-org/ovn-kubernetes/go-controller/pkg/crd/userdefinednetwork/v1/apis/informers/externalversions/userdefinednetwork/v1"
 
+	"github.com/ovn-org/ovn-kubernetes/go-controller/pkg/clustermanager/userdefinednetwork/template"
 	"github.com/ovn-org/ovn-kubernetes/go-controller/pkg/config"
 	"github.com/ovn-org/ovn-kubernetes/go-controller/pkg/factory"
 	"github.com/ovn-org/ovn-kubernetes/go-controller/pkg/util"
 )
 
 var _ = Describe("User Defined Network Controller", func() {
 	var (
-		udnClient   *udnfakeclient.Clientset
-		nadClient   *netv1fakeclientset.Clientset
-		udnInformer udninformer.UserDefinedNetworkInformer
-		nadInformer netv1Informer.NetworkAttachmentDefinitionInformer
-		kubeClient  *fake.Clientset
-		podInformer corev1informer.PodInformer
+		udnClient         *udnfakeclient.Clientset
+		nadClient         *netv1fakeclientset.Clientset
+		udnInformer       udninformer.UserDefinedNetworkInformer
+		cudnInformer      udninformer.ClusterUserDefinedNetworkInformer
+		nadInformer       netv1Informer.NetworkAttachmentDefinitionInformer
+		kubeClient        *fake.Clientset
+		podInformer       corev1informer.PodInformer
+		namespaceInformer corev1informer.NamespaceInformer
 	)
 
 	BeforeEach(func() {
 		udnClient = udnfakeclient.NewSimpleClientset()
 		udnInformer = udninformerfactory.NewSharedInformerFactory(udnClient, 15).K8s().V1().UserDefinedNetworks()
+		cudnInformer = udninformerfactory.NewSharedInformerFactory(udnClient, 15).K8s().V1().ClusterUserDefinedNetworks()
 		nadClient = netv1fakeclientset.NewSimpleClientset()
 		nadInformer = netv1informerfactory.NewSharedInformerFactory(nadClient, 15).K8sCniCncfIo().V1().NetworkAttachmentDefinitions()
 
 		kubeClient = fake.NewSimpleClientset()
 		sharedInformer := informerfactory.NewSharedInformerFactoryWithOptions(kubeClient, 15)
 		podInformer = sharedInformer.Core().V1().Pods()
+		namespaceInformer = sharedInformer.Core().V1().Namespaces()
 	})
 
+	newTestController := func(renderNADStub RenderNetAttachDefManifest) *Controller {
```

Comment:
and we need a test for this, that will create a NAD owned by CUDN in a namespace that is not matched by CUDN anymore, then start controller and make sure that stale NAD is deleted

Reply:
Some tests inject errors using fake clients PrependReactor, since util.GetOVNClientset does not expose the Fake entitiy its impossible to inject error using it.
Unfortunalty using util.GetOVNClientset in all test would not work

> and we need a test for this, that will create a NAD owned by CUDN in a namespace that is not matched by CUDN anymore, then start controller and make sure that stale NAD is deleted

Done
Covered by `should delete CUDN NAD in namespaces that no longer applies to namespace-selector` test

Reply:
> We should build namespaceTracker on controller startup with existing objects, so that we can do a cleanup if some NADs are stale now. To do so, you can use controller.StartWithInitialSync function.

this one is still not fixed. Test scenario should cover namespace label change while ovn-k was "off". So on startup you will see namespace and its NAD, but namespace should not match CUDN selector anymore, and its NAD should be removed.

Reply:
In order to support this scenario its necessary to implement some initial-sync logic for the namespace-tracker, so when the ovn-k starts it will populate the namespace-tacker with all managed NADs and reconcile them.
As a result, it will delete NADs in namespaces that no longer match the spec.

Should we do that in a follow-up PR?

Reply:
I think it is a small enough fix to add here, as the controller doesn't work properly without it anyway